Allow Asynchronously responding to Invoice Request

shaavan opened the pull request:
Introduce a new event and set of supporting functions to allow flexibility in responding asynchronously to a received invoice request.

This PR provides foundational work for further improvements in future. For example, this allows supporting Offers created in currency denomination, as a user can do appropriate pre-processing for a received InvoiceRequest, before sending an Invoice for it.

jkczyz commented on lightning/src/offers/invoice.rs:
```diff
@@ -340,18 +340,29 @@ macro_rules! invoice_builder_methods { (
+			return Ok(amount_msats);
+		}
+
+		if let Some(custom_amount_msats) = custom_amount_msats {
+			return Ok(custom_amount_msats);
+		}
```
It should probably be an error if we reach here and the offer has an amount denominated in bitcoin. That is the custom amount should only be applicable for currency-denominated offers. Not offers either denominated in bitcoin or with no amount. Though I _think_ in the latter case we'll fail to parse the invoice request. May want to double check that.

shaavan replied:
Makes sense!
Updated the check to make sure custom amount is provided only when an `Offer` is in Currency denomination and `InvoiceRequest`'s amount is absent in [pr3318.03](https://github.com/shaavan/rust-lightning/commits/pr3318.03)!

Also, I [confirmed](https://github.com/lightningdevkit/rust-lightning/blob/66fb5200986e0c145a16d74b04e471d9cdd08eb5/lightning/src/offers/invoice_request.rs#L1170-L1172) that we would fail to parse the `InvoiceRequest` if both offer's amount and IR's amount is absent.

Thanks a lot for the pointer!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4358,6 +4358,111 @@ where
+		let secp_ctx = &self.secp_ctx;
+		let expanded_key = &self.inbound_payment_key;
+
+		// Make sure that invoice_request amount and custom amount are not present at the same time.
+		if invoice_request.amount().is_some() && custom_amount_msats.is_some() {
+			return Err(Bolt12ResponseError::UnexpectedAmount)
+		}
```
Is this redundant with the check in the builder?

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4358,6 +4358,111 @@ where
+	}
+
+	fn get_response_for_verified_invoice_request(&self, invoice_request: &VerifiedInvoiceRequest, custom_amount_msats: Option<u64>) -> Result<(OffersMessage, PaymentHash), Bolt12ResponseError> {
+		let amount_msats = match InvoiceBuilder::<DerivedSigningPubkey>::amount_msats(
+			&invoice_request.inner, custom_amount_msats
+		) {
+			Ok(amount_msats) => amount_msats,
+			Err(error) => return Err(Bolt12ResponseError::SemanticError(error)),
+		};
```
Isn't this also called in the builder?

shaavan replied:
Yes!
But AFAIU, this is also called here so that the `amount_msats` value can be used with the `create_inbound_payment` and `create_blinded_payment_paths` functions.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -11038,6 +11038,14 @@ where
 
 		match message {
 			OffersMessage::InvoiceRequest(invoice_request) => {
+				if self.default_configuration.manually_handle_bolt12_messages {
+					let event = Event::InvoiceRequestReceived {
+						invoice_request, context, responder,
+					};
+					self.pending_events.lock().unwrap().push_back((event, None));
+					return None;
+				}
```
We should only create an event if the invoice request can be verified.

shaavan replied:
Makes sense!
Addressed the point in [pr3318.03](https://github.com/shaavan/rust-lightning/commits/pr3318.03)
Thanks a lot for the pointer!

shaavan commented:
Updated from [pr3318.01](https://github.com/shaavan/rust-lightning/commits/pr3318.01) to [pr3318.02](https://github.com/shaavan/rust-lightning/commits/pr3318.02) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3318.01..pr3318.02)):

Changes:
1. Rebase on main to resolve merge conflicts and fix CI.

shaavan commented:
Updated from [pr3318.02](https://github.com/shaavan/rust-lightning/commits/pr3318.02) to [pr3318.03](https://github.com/shaavan/rust-lightning/commits/pr3318.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr3318.02..pr3318.03)):
Addressed @jkczyz comments

Changes:
1. Update Responder to be required field.
2. Update InvoiceRequestReceived Docs.
3. Update condition check when the Custom Amount should be provided.
4. Create the InvoiceRequestReceived event only when InvoiceRequest is verified.

shaavan commented on lightning/src/events/mod.rs:
```diff
@@ -820,6 +821,23 @@ pub enum Event {
 		addresses: Vec<msgs::SocketAddress>,
 	},
+
+	/// Event triggered when manual handling is enabled, and an invoice request is received.
```
Updated docs in [pr3318.03](https://github.com/shaavan/rust-lightning/commits/pr3318.03)
Thanks a lot for the pointer! 🌟

shaavan commented on lightning/src/events/mod.rs:
```diff
@@ -820,6 +821,23 @@ pub enum Event {
+		/// `None` if the invoice wasn't sent with a reply path.
+		///
+		/// [`InvoiceError`]: crate::offers::invoice_error::InvoiceError
+		responder: Option<Responder>,
```
Update responder to be a required field in [pr3318.03](https://github.com/shaavan/rust-lightning/commits/pr3318.03)
Thanks for the suggestion!

jkczyz commented:
> I'm kinda skeptical of the approach of pushing yet more message handling out through the event handling interface. Looking back I'm really not a fan of the approach we took with `manually_handle_bolt12_invoices`, too. Instead of hooking things in the middle of their logic in `ChannelManager` and pushing them up as events, should we instead refactor the BOLT 12 handling in `ChannelManager` to expose an `OffersMessageHandler` which does the context testing and message validation and then passes the messages off to a handler interface.
> 
> `ChannelManager` can handle use this internally to do the first layer of message processing. For `invoice_request`s this basically leaves `ChannelManager` just fetching a blinded path and calling respond. For `invoice`s `ChannelManager` would basically just take the invoice and pass it to the outbound payment logic.

I'll need a more concrete description of this to make sure that I understand what you are proposing.

Let's call the refactored out struct `OffersMessageFlow` to avoid overloading `OffersMessageHandler`, which I presume `ChannelManger` would still implement and thus would typically still be used as a parameter to `OnionMessenger`. Then `OffersMessageFlow` would itself be parameterized by some new trait for users to implement, IIUC, which would need to trickle up to `ChannelManager` since it would be holding an `OffersMessageFlow`.

So, in practice, users would simply implement a small trait and parameterize `ChannelManager` with it. And implementations of this trait would be used to examine the messages and returning whether they should be acted on and possibly with what values (e.g., an amount in case of fiat-denominations). 

IIUC, the logic for constructing an invoice would still reside in `ChannelManager`, too, as constructing an invoice relies on many of its fields.

Does all of that sound right?

> I also regret not doing it this way initially, since I assume this would have reduced the code in LNDK rather nontrivially :/

Hmm... does it make it that much simpler? The above can be accomplished without `OffersMessageFlow` by simply adding another parameter to `ChannelManager`. Unless I misunderstood your proposal, the only thing `OffersMessageFlow` would used for would be the following two chunks of code:

https://github.com/lightningdevkit/rust-lightning/blob/a952d2d3d3bf70f9ddca46184a2fcbcab69aaa33/lightning/src/ln/channelmanager.rs#L11138-L11155

https://github.com/lightningdevkit/rust-lightning/blob/a952d2d3d3bf70f9ddca46184a2fcbcab69aaa33/lightning/src/ln/channelmanager.rs#L11240-L11243

Or do you also want to move the `Bolt12Invoice` construction into `OffersMessageFlow`? If so, `ChannelManager` would first call it to verify the request. Then depending on the result, call another method with a `payment_hash` and `payment_paths`.

Let me know if I misunderstood anything.

jkczyz commented:
> We could have the `OffersMessageFlow` be the only implementer of `OffersMessageHandler` and make users configure the `OnionMessenger` with an `OffersMessageFlow<ChannelManager>`, but we could also expose the guts of `OffersMessageFlow` crate-internal so that we can create this intermediation trait on the fly and pass it in to individual message handlers.

We also have `ChannelManager::pending_offers_messages` for user-initiated messages (e.g., used by `pay_for_offer` to enqueue invoice requests), which `OnionMessenger` queries. So if `OffersMessageFlow` implements `OffersMessageHandler`, we'd likely need to move those methods there. FWIW, it seems like we are (at least partially) recreating the old `InvoicePayer` abstraction since `OffersMessageFlow` would need to be parameterized by some trait that `ChannelManager` would implement. The `MessageRouter` trait probably could move to `OffersMessageFlow`, though I'm not sure if other/future onion-message uses would still remain in `ChannelManager`.

I'm actually more in favor of this now, especially if allows us to reduce the amount of code in `ChannelManager`.

jkczyz commented:
> I'm actually more in favor of this now, especially if allows us to reduce the amount of code in `ChannelManager`.

`InvoiceRequest` retries will be difficult, though. 😕

jkczyz commented:
@TheBlueMatt I was thinking about this a bit more. For the Fedimint case, the gateway will need to communicate with the client before responding or possibly (depending on the ultimate approach) wait for the federation to sign, IIUC. That would mean we'd either have to block waiting for the response or expose an asynchronous interface after all. This probably could be avoided for the currency use case if exchange rates are cached, though.

We could still refactor but seems we'd need to allow for asynchronous responses?

shaavan commented:
#3525 introduces `Bolt12Assessor`, allowing for custom logic to assess `InvoiceRequest` before handling, which serves as an alternative to this PR. Given this, I am closing this PR in favor of #3525.
